File: Weekly/week_07/dvHuni/OddEvenLinkedList.swift

import Foundation

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */

public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init() { self.val = 0; self.next = nil; }
    public init(_ val: Int) { self.val = val; self.next = nil; }
    public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
}

class Solution {
    func oddEvenList(_ head: ListNode?) -> ListNode? {
        var odd = head
        var even = head?.next
        let evenHead = even
        
        while even?.next != nil {
            odd?.next = even?.next
            even?.next = odd?.next?.next
            
            odd = odd?.next
            even = even?.next
        }
        
        odd?.next = evenHead
        
        return head
    }
}

File: Weekly/week_07/Keepbang/OddEvenLinkedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var oddEvenList = function(head) {

    if(head == null){
        return null;
    }

    let oddList = new ListNode(head.val,null);
    let oddHead = oddList;
    head = head.next;

    let evenList = null;
    let evenHead = evenList;
    if(head != null){
        evenList = new ListNode(head.val,null);
        evenHead = evenList;
        head = head.next;
    }

    let curState = -1;  //-1 : odd, 1 : even


    while(head != null){
        let tmp = new ListNode(head.val,null)
        if(curState == -1){
            oddList.next = tmp;
            oddList = oddList.next;
        }else{
            evenList.next = tmp;
            evenList = evenList.next;
        }
        curState *= -1;
        head = head.next;
    }

    oddList.next = evenHead;

    return oddHead;

};